### ISSUE TITLE
How to handle missing types from sub-graph plugin?

### ISSUE THREAD

**Issue by knpwrs** [hayes/pothos#478]
I have a type definition like the following:

```ts
export const Foo = builder.prismaObject('Foo', {
  fields: (t) => ({
    // ...
    bars: t.relation('bars', { subGraphs: ['Internal'] }),
  }),
});

builder.prismaObject('Bar', {
  subGraphs: ['Internal'],
  fields: (t) => ({
    // ...
  }),
});
```

When I try to create a schema with the public sub graph, I get the following runtime error:

```
Error: Bar (referenced by BarsConnectionEdge.node) does not exist in subGraph (Public)
```

I would have expected `BarsConnectionEdge.node` to inherit the subgraph configuration from the `t.relation` call, but this doesn't appear to be happening. I have the following configuration for the sub-graph plugin:

```ts
  subGraphs: {
    defaultForTypes: ['Public', 'Internal'],
    fieldsInheritFromTypes: true,
  },
```

The documentation says that missing types may occur, but it doesn't mention if there is any way to get around this. Is there a way to get around the issue for `t.relation` in particular? Can the behavior of `t.relation` be modified for the generated edge/node types to inherit the subgraph configuration?

**Comment by hayes** [comment]
I'm not sure I understand the question here.  It looks like you are explicitly saying that the Bar type is only part of the internal subgraph, and then defining a connection returning that type to be part of the public schema.   What types/fields are you expecting to be part of the public subgraph?

**Comment by knpwrs** [comment]
I'm trying to make an internal subgraph but the public subgraph still appears to be referencing something from the internal subgraph, seemingly through a type generated by `t.relation` (I didn't create `BarsConnectionEdge.node`, that was created for me).

This call works as expected:

```ts
export const internal = builder.toSchema({ subGraph: ['Public', 'Internal'] });
```

This call, which I expect to create my public schema, does not work and produces the error above.

```ts
export const publik = builder.toSchema({ subGraph: ['Public']] });
```

**Comment by hayes** [comment]
Maybe this will help: https://stackblitz.com/edit/typescript-munevt?file=index.ts

I think somewhere you have a `connection` field that might have the correct subGraph options, but the Connection object and Edge object created automatically might not.  You can add options for these objects by passing in additional options to the connection field.

**Comment by hayes** [comment]
It would probably be useful to add an option that automatically removes any fields that reference a type that has been removed.  Might also be worth updating the relay plugin so it is aware of how the subgraph plugin works and automatically extends the subGraph options for a connection field to the Connection and Edge types.  This is a place where the integration between plugins isn't very smart.

**Comment by knpwrs** [comment]
> It would probably be useful to add an option that automatically removes any fields that reference a type that has been removed.

I think that makes more sense than changing the relay plugin to be subgraph-aware. That would break encapsulation and be hard to maintain I'd think.

**Comment by hayes** [comment]
The thing that's weird with that solution is if you have a private `Thing` type, and a `thingsConnection` you end up with a `ThingsConnection` and `ThingsEdge`, if we remove fields that reference `Thing` that makes the schema work, but you still have the `ThingsConnection` and `ThingsEdge` types in the schema which isn't ideal.   I think maybe 2 options could solve this, adding an option to remove fields for types that are also removed (this might be something I can add without an option as the default behavior), and then also adding an option to strip out orphaned types (eg types that are not referenced by any fields in the schema).  The second part I think needs to be added as an option because it would be a breaking change

**Comment by knpwrs** [comment]
I just ran into this issue again with the Errors plugin. Since the Errors plugin creates its own types it is fundamentally incompatible with the subgraphs plugin.

**Comment by hayes** [comment]
Is this because the Error types are not removed from the public graph, or is there an additional issue?

**Comment by hayes** [comment]
I probably won't have time to fix this until Sunday, but it's fairly high on my list of priorities.  I think the be fix is pretty straightforward

**Comment by knpwrs** [comment]
> Is this because the Error types are not removed from the public graph, or is there an additional issue?

Here is a small example:

```ts
builder.queryType({
  subGraphs: ['Public', 'Internal'],
  fields: (t) => ({
    me: t.prismaField({
      type: 'AppUser',
      subGraphs: ['Public'],
      authScopes: {
        authenticated: true,
      },
      errors: { types: [ForbiddenError] },
      resolve: async (_query, _root, _args, _ctx, _info) => {
        // snip
      },
    }),
  }),
});
```

And this is the error I get:

```
Error: QueryMeResult (referenced by Query.me) does not exist in subGraph (Public)
```

> I probably won't have time to fix this until Sunday, but it's fairly high on my list of priorities. I think the be fix is pretty straightforward

I might be able to put something together at some point but by no means wait for me :)

**Comment by knpwrs** [comment]
Subgraphs also appear to cause issues with the with-input plugin:

```
Property 'input' is missing in type 'InputShapeFromFields<{ input: InputFieldRef<InputShapeFromFields< // etc
```

**Comment by hayes** [comment]
Just pushed a fix for this, release should be published in a few minutes.  Let me know if you are still running into any issues after upgrading!

**Comment by knpwrs** [comment]
At first glance things appear to be working, thank you!

Just curious -- what changed that there didn't need to be an option anymore? I didn't change anything other than upgrading my dependencies.

**Comment by hayes** [comment]
The changes should only affect APIs that couldn't be compiled anyways, so while there is technically potential for a breaking change, I think it should be fine for all real world usage.  This should provide a better experience for all users.  If you run into any issues with it let me know.

**Comment by knpwrs** [comment]
So after trying this out more thoroughly I'm still hitting an issue. The server doesn't crash anymore (yay!) but there are fields missing from the schema now for plugin-generated types.

Consider the `me` example from earlier:

```ts
builder.queryType({
  subGraphs: ['Public', 'Internal'],
  fields: (t) => ({
    me: t.prismaField({
      type: 'AppUser',
      subGraphs: ['Public'],
      authScopes: {
        authenticated: true,
      },
      errors: { types: [ForbiddenError] },
      resolve: async (_query, _root, _args, _ctx, _info) => {
        // snip
      },
    }),
  }),
});
```

Now instead of an error, `me` is entirely excluded from the `Public` schema. It is entirely missing from the GraphiQL explorer. I also get the following error from graphql code generator:

```
Error 1: GraphQLDocumentError: Unknown type "QueryMeSuccess".
     at /home/node/app/app/routes/index.tsx:5:16
```

I can work around this by defaulting types to _all_ subgraphs and then selectively overriding individual types to only certain subgraphs, which results in their generated types being dropped from the schema. I would argue, however, that it is more robust to be able to define a smaller subset of the schemas as default and then selectively add schemas. Defaulting to `['Public', 'Internal']` makes it possible to accidentally leak types to the public schema if you forget to configure them, but making the default `['Internal']` makes it impossible for the generated types to show up in the public schema, and setting the defaults to `[]` / not defining any defaults makes it impossible for them to show up in any schema, and when the generated type doesn't appear in the schema the parent type is dropped as well.

Maybe an option like `generatedTypesInheritFromParentTypes` is in order? There would need to be some mechanism in place for the subgraph plugin to know that a type is generated and what type generated it.

**Comment by hayes** [comment]
That makes complete sense... This is easy to work around for the relay plugin:

inside realyOptions in the builder, you can add options for connection objects, edge objects, pageInfo, etc.  Setting the subGraphs and fieldsInheritFromType: true for all of these will get the relay connection objects to show up for fields included in the subGraph.

For the errors plugin its more complicated.  You can currently pass options for each field that uses the errors plugin.  doing something like: `errors: { result: { subGraphs: ['Public'] }, union: { subGraphs: ['Public'], types: [...] }` is pretty verbose though.  I think having builder level default options for error unions and result objects would work.

> Maybe an option like generatedTypesInheritFromParentTypes is in order? There would need to be some mechanism in place for the subgraph plugin to know that a type is generated and what type generated it.

This is complicated for multiple reasons.  The mechanism for knowing what generated types doesn't currently exist.  Types can also be used by multiple fields even when generated, and if those fields use different subGraphs it wouldn't be obvious which subFields to use. 

Another alternative, that might also be generally useful would be to provide a way to pattern match types into subGraphs:

```ts
defaultSubGraphsForTypes: {
   '*': ['Internal'],
   '*Connection': ['Public'],
   'Admin*': ['Internal'],
}
 ```

**Comment by hayes** [comment]
I'll try to put something together in the next couple days.  Hopefully the workaround mentioned above will work for you in the mean time

**Comment by hayes** [comment]
Just pushed a change with a few more options for configuring defaults for generated types.  This doesn't really solve the underlying problem, but helps with the workaround:

```ts
export default new SchemaBuilder<Types>({
  plugins: [ErrorsPlugin, SubGraphPlugin, RelayPlugin, WithInputPlugin],
  subGraphs: {
    fieldsInheritFromTypes: true,
    defaultForTypes: ['Private'],
  },
  errorOptions: {
    defaultTypes: [],
    defaultUnionOptions: {
      subGraphs: ['Private', 'Public'],
    },
    defaultResultOptions: {
      subGraphs: ['Private', 'Public'],
      defaultSubGraphsForFields: ['Private', 'Public'],
    },
  },
  relayOptions: {
    clientMutationId: 'omit',
    cursorType: 'String',
    defaultConnectionTypeOptions: {
      subGraphs: ['Private', 'Public'],
      defaultSubGraphsForFields: ['Private', 'Public'],
    },
    defaultEdgeTypeOptions: {
      subGraphs: ['Private', 'Public'],
      defaultSubGraphsForFields: ['Private', 'Public'],
    },
    pageInfoTypeOptions: {
      subGraphs: ['Private', 'Public'],
      defaultSubGraphsForFields: ['Private', 'Public'],
    },
    defaultPayloadTypeOptions: {
      subGraphs: ['Private', 'Public'],
      defaultSubGraphsForFields: ['Private', 'Public'],
    },
    defaultMutationInputTypeOptions: {
      subGraphs: ['Private', 'Public'],
    },
  },
  withInput: {
    typeOptions: { subGraphs: ['Private', 'Public'] },
  },
});
```

**Comment by knpwrs** [comment]
> Hopefully the workaround mentioned above will work for you in the meantime

Absolutely, it will work for the time being, especially with those changes you just pushed. Thank you, again, very much!

**Comment by knpwrs** [comment]
I'm seeing a strange issue now with SubGraphs and Interfaces now. It seems like the concrete types don't make it into any subgraph. Here is an entire running example in one file:

```
import { createServer } from '@graphql-yoga/node';
import SchemaBuilder from '@pothos/core';
import SimpleObjectsPlugin from '@pothos/plugin-simple-objects'
import SubGraphPlugin from '@pothos/plugin-sub-graph'

const builder = new SchemaBuilder<{
  SubGraphs: 'Public' | 'Internal';
}>({
  plugins: [
    SimpleObjectsPlugin,
    SubGraphPlugin,
  ],
  subGraphs: {
    fieldsInheritFromTypes: true,
    defaultForTypes: ['Internal'],
  },
});

const ContactInfo = builder.simpleObject('ContactInfo', {
  subGraphs: ['Public', 'Internal'],
  fields: (t) => ({
    email: t.string({
      nullable: false,
    }),
    phoneNUmber: t.string({
      nullable: true,
    }),
  }),
});

const Node = builder.simpleInterface('Node', {
  subGraphs: ['Public', 'Internal'],
  fields: (t) => ({
    id: t.id({
      nullable: false,
    }),
  }),
});

builder.simpleObject('User', {
  subGraphs: ['Public', 'Internal'],
  interfaces: [Node],
  fields: (t) => ({
    firstName: t.string(),
    lastName: t.string(),
    contactInfo: t.field({
      type: ContactInfo,
      nullable: false,
    }),
  }),
});

builder.queryType({
  subGraphs: ['Public', 'Internal'],
  fields: (t) => ({
    node: t.field({
      type: Node,
      args: {
        id: t.arg.id({ required: true }),
      },
      resolve: (parent, args, { User }) => {
        return {
          id: '1003',
          firstName: 'Leia',
          lastName: 'Organa',
          contactInfo: {
            email: 'leia@example.com',
            phoneNUmber: null,
          },
        };
      },
    }),
  }),
});

const server = createServer({
  schema: builder.toSchema({ subGraph: ['Internal'] }),
  port: 9999,
});

server.start();
```

When I visit `http://localhost:3000/graphql`, I see the following:

<img width="570" alt="Screen Shot 2022-08-25 at 5 20 13 AM" src="https://user-images.githubusercontent.com/174864/186640108-cd91f5f3-a2b1-4c80-bbb4-57d4cdd6322c.png">

However, when I comment out the `SubGraphPlugin`:

```diff
diff --git a/src/index.ts b/src/index.ts
index 854b0ef..bc01c2c 100644
--- a/src/index.ts
+++ b/src/index.ts
@@ -8,7 +8,7 @@ const builder = new SchemaBuilder<{
 }>({
   plugins: [
     SimpleObjectsPlugin,
-    SubGraphPlugin,
+    // SubGraphPlugin,
   ],
   subGraphs: {
     fieldsInheritFromTypes: true,
```

And then visit `http://localhost:3000/graphql` again, I see the following:

<img width="577" alt="image" src="https://user-images.githubusercontent.com/174864/186640942-d8e21709-b372-476e-9c98-9b9d2e08f43f.png">

I've tried changing the order of the plugins and checking all my usages of `subGraphs` but it still doesn't appear to be working.

**Comment by hayes** [comment]
Pushed a fix here: https://github.com/hayes/pothos/commit/95fcbaf8102524c2dbb03a90ade5f3ca99191026

**Comment by hayes** [comment]
Fix is published in 3.10.0

**Comment by knpwrs** [comment]
You're amazing, thank you so much!
